### ISSUE TITLE
Assertion failed: _lastFuture == null

### ISSUE THREAD

**Issue by Zekfad** [rrousselGit/riverpod#2041]
**Describe the bug**
Using following chain of providers
1. AsyncNotifier
2. **AsyncNotifier**: .future from 1st provider
3. AsyncNotifier: .future from 2nd provider

Will cause following error:
<details>
<summary> Log </summary>

```
══╡ LOGGER RECORD / SEVERE: 2022-12-30 20:05:04.040 ╞═══════════════════════════════════════════════
ErrorText unknown error.
Assertion failed: file:///C:/Users/vorob/AppData/Local/Pub/Cache/hosted/pub.dev/riverpod-2.1.3/lib/src/async_notifier/base.dart:333:18
_lastFuture == null
"bad state"
<Debug frame / internal> <unknown>:<unknown>/<unknown>:288:49 throw_ (1 frame)
<Debug frame / internal> <unknown>:<unknown>/<unknown>:29:3 assertFailed (1 frame)
package:riverpod/src/async_notifier/base.dart:333:30 <fn> (1 frame)
package:riverpod/src/async_notifier/base.dart:304:11 <fn> (1 frame)
package:riverpod/src/async_notifier/base.dart:325:47 [_handleAsync] (1 frame)
package:riverpod/src/async_notifier/base.dart:271:5 handleFuture (1 frame)
package:riverpod/src/async_notifier/base.dart:450:9 <fn> (1 frame)
package:riverpod/src/result.dart:68:16 when (1 frame)
package:riverpod/src/async_notifier/base.dart:445:19 create (1 frame)
package:riverpod/src/framework/element.dart:418:7 buildState (1 frame)
package:riverpod/src/framework/element.dart:356:5 [_performBuild] (1 frame)
package:riverpod/src/framework/element.dart:319:7 flush (1 frame)
package:riverpod/src/framework/scheduler.dart:62:40 [_performRefresh] (1 frame)
package:riverpod/src/framework/scheduler.dart:50:5 [_task] (1 frame)
package:flutter_riverpod/src/framework.dart:392:12 build (1 frame)
<Debug frame / flutter internal> package:flutter/src/widgets/framework.dart:4995:15 performRebuild (1 frame)
<Debug frame / flutter internal> package:flutter/src/widgets/framework.dart:4715:7 rebuild (1 frame)
<Debug frame / flutter internal> package:flutter/src/widgets/framework.dart:2767:18 buildScope (1 frame)
<Debug frame / flutter internal> package:flutter/src/widgets/binding.dart:864:9 drawFrame (1 frame)
<Debug frame / flutter internal> package:flutter/src/rendering/binding.dart:381:5 [_handlePersistentFrameCallback] (1 frame)
<Debug frame / flutter internal> package:flutter/src/scheduler/binding.dart:1289:15 [_invokeFrameCallback] (1 frame)
<Debug frame / flutter internal> package:flutter/src/scheduler/binding.dart:1218:9 handleDrawFrame (1 frame)
<Debug frame / flutter internal> package:flutter/src/scheduler/binding.dart:1076:5 [_handleDrawFrame] (1 frame)
<Debug frame / internal> <unknown>:<unknown>/<unknown>:1193:13 invoke (1 frame)
<Debug frame / internal> <unknown>:<unknown>/<unknown>:220:5 invokeOnDrawFrame (1 frame)
<Debug frame / internal> <unknown>:<unknown>/<unknown>:190:45 <fn> (1 frame)
<Debug frame / internal> <unknown>:<unknown>/<unknown>:367:37 _checkAndCall (1 frame)
<Debug frame / internal> <unknown>:<unknown>/<unknown>:372:39 dcall (1 frame)
════════════════════════════════════════════════════════════════════════════════ 1672419904040000 ══
```
</details>

**To Reproduce**

https://dartpad.dev/?id=b54462f0563ceb81ac6697c24bd79a41

**Expected behavior**

No errors.

**Comment by BreX900** [comment]
I add more information

**Describe the bug**
When you invalidate 2 providers linked together the state of the FutureProvider/StreamProvider.future is invalid.
This mainly happens when:
- The function is declared async without waiting for any future
- By invalidating 2 providers at the same time
- By connecting the 2 providers to each other with watch

```
I/flutter (22874): Disposed!
I/flutter (22874): Disposed!
I/flutter (22874): AsyncError<DateTime>(value: 2022-12-30 19:18:40.164640, error: 'package:riverpod/src/async_notifier/base.dart': Failed assertion: line 333 pos 18: '_lastFuture == null': bad state, stackTrace: #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
I/flutter (22874): #1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
I/flutter (22874): #2      FutureHandlerProviderElementMixin._handleAsync.<anonymous closure> (package:riverpod/src/async_notifier/base.dart:333:18)
I/flutter (22874): #3      FutureHandlerProviderElementMixin.handleFuture.<anonymous closure> (package:riverpod/src/async_notifier/base.dart:304:11)
I/flutter (22874): #4      FutureHandlerProviderElementMixin._handleAsync (package:riverpod/src/async_notifier/base.dart:325:47)
I/flutter (22874): #5      FutureHandlerProviderElementMixin.handleFuture (package:riverpod/src/async_notifier/base.dart:271:5)
I/flutter (22874): #6      FutureProviderElement.create (package:riverpod/src/future_provider/base.dart:84:5)
I/flutter (22874): #7      ProviderElementBase.buildState (package:riverpod/src/framework/element.dart:418:7)
I/flutter (22874): #8      Provider
```

**To Reproduce**

```dart
final provider2 = FutureProvider((ref) async {
  return DateTime.now();
});

final provider1 = FutureProvider((ref) async {
  ref.onDispose(() => print('Disposed!'));

  await ref.watchFuture(provider2);

  return DateTime.now();
});
```
```dart
class _Home extends ConsumerWidget {
  const _Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    ref.watch(provider1);

    return Scaffold(
      body: InkWell(
        onTap: () {
          ref.invalidate(provider1);
          ref.invalidate(provider2);
        },
        child: const SizedBox.expand(),
      ),
    );
  }
}
```
**Expected behavior**
I shouldn't get any errors and provider1 shouldn't call onDispose 2 times.


**One more things**
I can imagine the problem is provider1 gets destroyed more than once and recreated more than once.

**Comment by SimoneBressan** [comment]
@rrousselGit This happens:
A provider `provider1` listens for a "future" provider `provider2`.
- When you invalidate both providers this happens: all providers lose parental knowledge.
- Scheduling starts.
- Start flush `provider1`. He will try to rebuild the fathers but he has none. It rebuilds itself and listens to the `provider2` of its future which is already completed.
- Start building `provider2`. Will update future invalidating children specifically `provider1` is invalidated.
- New scheduling starts.
- Start building `provider1`. It rebuilds itself and listens to the `provider2` of its new future which is not yet completed.

This solved the problem. By nullifying `_futureCompleter`.
```dart
mixin FutureHandlerProviderElementMixin<T> on ProviderElementBase<AsyncValue<T>> {
  @override
  void runOnDispose() {
    // Stops listening to the previous async operation
    _lastFutureSub?.call();
    _lastFutureSub = null;
    _lastFuture = null;
    _cancelSubscription?.call();
    _cancelSubscription = null;
    _futureCompleter?.future.ignore();
    _futureCompleter = null;
    super.runOnDispose();
  }
}
```

This solution works only when all providers of the tree to be invalidated are marked as invalidated. Otherwise, providers will risk listening to an "old" future.

**Comment by AhmedLSayed9** [comment]
I encounter the same issue when dealing with declarative pagination:
https://dartpad.dev/?id=05d93f8e8c71368eab0c03fae4f77d79
As you can see in my sample, `current item` watch `prev item` which leads to throwing `_lastFuture == null` when refreshing all pages.

**Comment by yardwerkz** [comment]
I have this same problem. Is there a workaround or fix?

Thanks,
Hadil

**Comment by rrousselGit** [comment]
Thanks for the reproduction samples. I'm on a fix and will keep you posted

**Comment by rrousselGit** [comment]
Minimal reproduction:

```dart
test('Does not throw assertion error in certain conditions', () async {
  // Regression test for https://github.com/rrousselGit/riverpod/issues/2041

  final container = createContainer();

  final testNotifierProvider = FutureProvider.autoDispose<int>((ref) => 0);
  final proxyProvider = FutureProvider.autoDispose<int>(
    (ref) => ref.watch(testNotifierProvider.future),
  );
  final testProvider = FutureProvider.autoDispose<int>(
    (ref) async => (await ref.watch(proxyProvider.future)) + 2,
  );

  container.listen<AsyncValue<void>>(
    testProvider,
    (previous, next) {
      if (!next.isLoading && next is AsyncError) {
        Zone.current.handleUncaughtError(next.error, next.stackTrace);
      }
    },
    fireImmediately: true,
  );

  container.invalidate(testNotifierProvider);
  container.invalidate(testProvider);

  await container.pump();
});
```




It appears that a workaround is to read the root provider after invalidating it.
Doing:
```dart
container.invalidate(testNotifierProvider);
container.invalidate(testProvider);
container.read(testProvider);
```

Does not throw somehow.

**Comment by rrousselGit** [comment]
Looks like providers are rebuilding in the incorrect order, and the leaf (`testProvider`) somehow tries to rebuild twice.
That's likely the root of the issue here. I'll investigate more.

**Comment by BreX900** [comment]
@rrousselGit 

I think the biggest problem is that the ref does not change with each new recreation of a provider. An invalidated provider can continue to use the ref even if it is from another build. This creates invalid states.
This always or very often occurs when you listen to future providers

Let me try to explain again: the ref should be linked to the life cycle of the provider's create/build value/future 

Right now: the ref is always valid for any create/build of a provider's value even if that build has been discarded 

For example
```dart

final xProvider = FuturePeovider((ref) {
  ref.watch(provider1.future) // 1. here you invalidate the provider before the future completes
  ref.watch(provider2.future)
})
```

2. Provider X is then recreated but the previous creation has not completed. When the future of provider 1 of the first creation completes you will have provider X listen twice to provider 2 and once to provider 1
This happens because the ref of the first creation is still valid 

I can't explain it well, if it's not clear let me know

**Comment by rrousselGit** [comment]
That's unrelated to the issue at hand.

I've been thinking about something similar, but it wouldn't work with the way Notifier is. Since the Notifier persists across watch updates, its ref needs to always be valid. That would render the idea of changing ref instance between rebuilds moot.

It may be something for v3.

**Comment by BreX900** [comment]
@rrousselGit This error also appeared to me for this reason. But yes it is not directly related to the example of this issue. Thanks

**Comment by rrousselGit** [comment]
This error is completely unrelated :)
It is most likely caused by a refactoring of modifiers, where `.future` is no-longer a provider but "only".

I've just checked and it appears that all modifiers are bugged. This includes .future, but also .select/.notifier

It just happens that this assert is the most common failure point

**Comment by rrousselGit** [comment]
This is gonna be difficult to fix without also implementing https://github.com/rrousselGit/riverpod/issues/1253, which is breaking.

A v3 is already on the way. I'll add this to the v3 branch along with #1253 and make a dev release.
V3 is mostly internal switcheroo. Upgrading shouldn't be an issue.

Considering how big of an internal refactoring is needed to fix this bug, I wouldn't be comfortable with shipping this in a non-breaking release. Even if rare, some folks out there do use the internals, so breaking is breaking :shrug:

**Comment by rrousselGit** [comment]
No yet no. I'll post it here when I make it.

Since it's been around for a while and requires a lot of work to fix, I've ended up working on docs instead for now.
The v3 release will come right after r

**Comment by lucavenir** [comment]
I stumbled on this exact problem, except I just have FutureProviders in my case. I was lucky enough I could just drop the .future getter and use a .requireValue instead with no issues.

**Comment by rrousselGit** [comment]
I started working on this again about a week ago.

I'm trying what I can to find a fix that's backward compatible with 2.0, but that's quite difficult.

**Comment by lucavenir** [comment]
So Riverpod 3 won't have this issue at all?

**Comment by rrousselGit** [comment]
Yes. What's unclear is whether the fix will land in 2.0 too.

I'd like to fix this without forcing people to migrate to higher versions. But it's a quite deep change apparently

**Comment by rrousselGit** [comment]
Ok the issue is in fact a lot different from what I originally though.
TL;DR the problem is that sometimes adding a listener on `provider.future` immediately notifies this listener ; when that shouldn't happen.

The double rebuild is just a consequence of that.

Better reproduction:

```dart
  test(
      '.future does not immediately notify listeners when adding a new listener '
      'to .future flushes the provider', () async {
    // Regression test for https://github.com/rrousselGit/riverpod/issues/2041

    final container = createContainer();
    final onFuture = Listener<Future<int>>();

    final dep = FutureProvider((ref) => 0);
    final provider = Provider(
      (ref) {
        ref.listen(dep.future, onFuture.call);
        return 0;
      },
    );

    container.read(dep);
    container.invalidate(dep);

    container.read(provider);

    verifyZeroInteractions(onFuture);
  }
```


Since the problem is much simpler than I originally thought, I'm hopeful that I'll be able to find a fix for 2.0

**Comment by rrousselGit** [comment]
Fixed as part of Riverpod 2.5.1
